### PR TITLE
refactor access type array to access flags

### PR DESCRIPTION
**Pull request by YunHsiao**
Re: cocos-creator/3d-tasks#

Changelog:
 * refactor barrier interface, got rid of the array constructs
 * renamed `GlobalBarrier` to `GeneralBarrier` for better clarity (it's now being used to describe attachment synchronizations too)
 * corrected 3.0.0 deprecation behavior: polyfills should have never depended on current definitions

### CODE REVIEW

File: cocos/core/gfx/deprecated-3.5.0.ts [hunk 1]
```diff
@@ -105,3 +106,32 @@ removeProperty(Feature, 'Feature', [
              + 'let isSupported = device.getFormatFeatures(Format.RGB8) !== FormatFeatureBit.NONE;',
     },
 ]);
+
+removeProperty(ColorAttachment.prototype, 'ColorAttachment', [
```
Comment (YunHsiao) [inline cocos/core/gfx/deprecated-3.5.0.ts:110]: Deprecation warnings for the breaking changes in this PR are implemented here

File: native/cocos/renderer/gfx-base/GFXDef-common.h [hunk 2]
```diff
@@ -764,29 +765,29 @@ using ClearFlags = ClearFlagBit;
...
-    uint32_t maxShaderStorageBlockSize{0U};
-    uint32_t maxUniformBufferBindings{0U};
-    uint32_t maxUniformBlockSize{0U};
-    uint32_t maxTextureSize{0U};
-    uint32_t maxCubeMapTextureSize{0U};
-    uint32_t uboOffsetAlignment{1U};
-
-    uint32_t maxComputeSharedMemorySize{0U};
-    uint32_t maxComputeWorkGroupInvocations{0U};
+    uint32_t maxVertexAttributes{0};
```
Comment (minggo) [inline native/cocos/renderer/gfx-base/GFXDef-common.h:781]: Will not cause clang-tidy error?

File: native/cocos/renderer/gfx-base/GFXDef-common.h [hunk 3]
```diff
@@ -1164,17 +1171,17 @@ struct RenderPassInfo {
...
 
-struct TextureBarrierInfo {
-    std::vector<AccessType> prevAccesses;
-    std::vector<AccessType> nextAccesses;
+struct ALIGNAS(8) TextureBarrierInfo {
+    AccessFlags prevAccesses{AccessFlagBit::NONE};
+    AccessFlags nextAccesses{AccessFlagBit::NONE};
 
-    bool discardContents{false};
+    uint64_t discardContents{0}; // @ts-boolean
```
Comment (minggo) [inline native/cocos/renderer/gfx-base/GFXDef-common.h:1191]: Why not use bool as TS?

File: native/cocos/renderer/gfx-base/GFXDef-common.h (hunk 2 above)
Comment (YunHsiao) [reply native/cocos/renderer/gfx-base/GFXDef-common.h:781]: Nope. Not sure why though

File: native/cocos/renderer/gfx-base/GFXDef-common.h (hunk 3 above)
Comment (YunHsiao) [reply native/cocos/renderer/gfx-base/GFXDef-common.h:1191]: The 64-bit type is used to make sure the memory layout is not implicitly padded.

Comment (minggo) [reply native/cocos/renderer/gfx-base/GFXDef-common.h:1191]: Got it, could you please add some comments for it?

File: native/cocos/renderer/gfx-base/GFXDef-common.h [hunk 4]
```diff
@@ -1212,48 +1219,45 @@ struct InputState {
     AttributeList attributes;
 };
 
-// Use uint32_t for all boolean values to convert memory to RasterizerState* in shared memory.
```
Comment (minggo) [inline native/cocos/renderer/gfx-base/GFXDef-common.h:1215]: Is there any reason to delete the comments?

Comment (YunHsiao) [reply native/cocos/renderer/gfx-base/GFXDef-common.h:1215]: I thought we don't have the shared memory system anymore

Comment (minggo) [reply native/cocos/renderer/gfx-base/GFXDef-common.h:1215]: It is needed for performance, as the states are checked and modified in 2D components.

**Review by minggo — APPROVED** [review]

File: native/cocos/renderer/gfx-base/GFXDef-common.h (hunk 4 above)
Comment (minggo) [reply native/cocos/renderer/gfx-base/GFXDef-common.h:1215]: The states are checked and modified many times every frame.

**Review by minggo — APPROVED** [review]

**Pull request merged**